File: LeetCode/DP/JS/Leetcode467-unique-substrings-in-wraparound-string.js
var findSubstringInWraproundString = function (p) {
    let mp = new Map();
    for (let i = 0; i < p.length;) {
        // console.log(i);
        let j = i + 1;
        while (j < p.length && (p[j].charCodeAt() - p[j - 1].charCodeAt() == 1 || (p[j] === 'a' && p[j - 1] === 'z'))) j++;
        // console.log(j);
        while (i < j) {
            mp.set(p[i], Math.max(mp.get(p[i]) || 0, j - i));
            i++;
        }
    }
    // console.log(mp);
    let res = 0;
    for (const it of mp.values()) res += it;
    return res;
};

File: LeetCode/DP/Rust/Leetcode467-unique-substrings-in-wraparound-string.rs
use std::collections::HashMap;
impl Solution {
  pub fn find_substring_in_wrapround_string(p: String) -> i32 {
    let mut mp = HashMap::new();
    let mut i = 0;
    let p = p.as_bytes();
    while i < p.len() {
      let mut j = i + 1;
      while j < p.len() && (p[j] - p[j - 1] == 1 || (p[j] == b'a' && p[j - 1] == b'z')) {
        j += 1;
      }
      while i < j {
        let cnt = mp.entry(&p[i..i + 1]).or_insert(j - i);
        *cnt = (*cnt).max(j - i);
        i += 1;
      }
    }
    let mut res = 0;
    for (_k, v) in mp {
      res += v;
    }
    res as i32
  }
}
